### PR TITLE
Dynamically shaped image support for `tf.image.resize_image_with_crop_or_pad`

### PR DESCRIPTION
**Pull request by gaohuazuo**
This patch is intended for issue https://github.com/tensorflow/tensorflow/issues/521

A new keyword argument `dynamic_shape=False` is added to these functions in `tensorflow.python.image_ops`:
`crop_to_bounding_box` `pad_to_bounding_box` `resize_image_with_crop_or_pad` `_ImageDimensions`

The change is fully backward compatible. The new functionality of `resize_image_with_crop_or_pad` is tested and worked well. However, the dynamic version of `crop_to_bounding_box` `pad_to_bounding_box` is missing many argument checks compared to the static version.

Keeping both a dynamic and a static version of the same function is quite ugly. Maybe we should keep only the dynamic version?

**TODO**
Test cases

### CODE REVIEW

File: tensorflow/python/ops/image_ops.py [hunk 1]
```diff
@@ -478,40 +492,46 @@ def crop_to_bounding_box(image, offset_height, offset_width, target_height,
...
 
   Raises:
     ValueError: If the shape of `image` is incompatible with the `offset_*` or
-    `target_*` arguments
+    `target_*` arguments, and `dynamic_shape` is set to `False`.
   """
   image = ops.convert_to_tensor(image, name='image')
-  _Check3DImage(image, require_static=True)
-  height, width, _ = _ImageDimensions(image)
+  _Check3DImage(image, require_static=(not dynamic_shape))
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: These not... are ugly. Can you change the argument from dynamic_shape=False to static_shape_check=True? Then all the functions have the same kind of parameter.

**Comment by martinwicke** [comment]
This is great. Can you add tests for this? I think having the ability to use static checks is useful (sometimes), and backwards compatibility is nice, so I don't think you should remove he static checks.

**Comment by gaohuazuo** [comment]
Should these functions simply skip checks that cannot be performed at graph construction time? That would eliminate the need  of a new argument.

**Comment by martinwicke** [comment]
That would mean that tensors of unknown size will pass all the static tests? That's a good idea. I think the danger is minimal here. The errors in the resize kernels are not too bad anyway.

File: tensorflow/python/ops/image_ops.py [hunk 2]
```diff
@@ -427,44 +433,57 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
...
-  if target_width < width:
-    raise ValueError('target_width must be >= width')
-  if target_height < height:
-    raise ValueError('target_height must be >= height')
-
   after_padding_width = target_width - offset_width - width
   after_padding_height = target_height - offset_height - height
 
-  if after_padding_width < 0:
+  if not isinstance(offset_height, ops.Tensor) and offset_height < 0:
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: Using `isinstance(something, ops.Tensor)` here is too clumsy.

Solution 1:
Use some helper function to check whether the value is a "literal" or "symbol".

Solution 2:
Make the framework automatically evaluate all ops that have constant inputs at graph construction time (like C++ constexpr).

Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: Solution 2 is not something that will happen right now (although constant folding does happen before execution)

How about making _is_tensor (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/python/framework/tensor_util.py#L218) public and using that?

**Comment by vrv** [comment]
@gaohuazuo you'll have to ping us when this is ready for review

**Comment by gaohuazuo** [comment]
@martinwicke Ready for review.

File: tensorflow/contrib/framework/python/framework/tensor_util.py [hunk 3]
```diff
@@ -215,8 +216,19 @@ def with_same_shape(expected_tensor, tensor):
     return with_shape(expected_shape, tensor)
 
 
-def _is_tensor(t):
-  return isinstance(t, (ops.Tensor, ops.SparseTensor, variables.Variable))
+def is_tensor(x):
+  """Check for tensor types.
+  If the object represents a tensor, returns True. Equivalent to
```
Comment (martinwicke) [inline tensorflow/contrib/framework/python/framework/tensor_util.py]: `True`

File: tensorflow/contrib/framework/python/framework/tensor_util.py [hunk 4]
```diff
@@ -215,8 +216,19 @@ def with_same_shape(expected_tensor, tensor):
...
+  `isinstance(x, [tf.Tensor, tf.SparseTensor, tf.Variable])`.
+
+  Args:
+    x: An python object to check.
+
+  Returns:
+    True if `x` represents a tensor, False if not.
```
Comment (martinwicke) [inline tensorflow/contrib/framework/python/framework/tensor_util.py]: `True`, `False`

File: tensorflow/python/ops/image_ops.py [hunk 5]
```diff
@@ -194,12 +195,18 @@ def _ImageDimensions(images):
     images: 4-D Tensor of shape [batch, height, width, channels]
 
   Returns:
-    list of integers [batch, height, width, channels]
+    list of integers [batch, height, width, channels], when static shape is
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: `[batch, height, width, channels]`

File: tensorflow/python/ops/image_ops.py [hunk 6]
```diff
@@ -194,12 +195,18 @@ def _ImageDimensions(images):
...
+    fully defined.
+    list of integer scalar tensors [batch, height, width, channels], when
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: here too

File: tensorflow/python/ops/image_ops.py [hunk 7]
```diff
@@ -427,36 +434,47 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
...
   after_padding_width = target_width - offset_width - width
   after_padding_height = target_height - offset_height - height
 
-  if after_padding_width < 0:
+  if not is_tensor(offset_height) and offset_height < 0:
+    raise ValueError('offset_height must be >= 0')
+  if not is_tensor(offset_width) and offset_width < 0:
+    raise ValueError('offset_width must be >= 0')
+
+  if (all(not is_tensor(i) for i in [target_width, width]) and
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: for sets of 2, is_tensor(target_width) and is_tensor(width) is much easier to read

File: tensorflow/python/ops/image_ops.py [hunk 8]
```diff
@@ -427,36 +434,47 @@ def pad_to_bounding_box(image, offset_height, offset_width, target_height,
...
+      target_width < width):
+    raise ValueError('target_width must be >= width')
+  if (all(not is_tensor(i) for i in [target_height, height]) and
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops.py]: here too

File: tensorflow/python/ops/image_ops_test.py [hunk 9]
```diff
@@ -551,21 +567,31 @@ def testError(self):
 class PadToBoundingBoxTest(test_util.TensorFlowTestCase):
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: Can you add a test that checks that it is possible to fail these various checks? I know that should have existed before but now we need it even more.

**Comment by martinwicke** [comment]
Thanks! Added bunch of comments. Mainly: Some doc nits, some nit nits, and, could you add another test that assertRaises appropriately if you pass in illegal values, and with also asserts that if you pass in tensors which have illegal values it still fails (although not with ValueError, but probably with OpError).

**Comment by gaohuazuo** [comment]
@martinwicke Test cases updated. Also added corresponding run time check for every static check, though not sure if these checks worth the cost of more complexity in the graph.

File: tensorflow/python/ops/image_ops.py [hunk 10]
```diff
@@ -212,16 +247,26 @@ def _Check3DImage(image, require_static=True):
 
   Raises:
     ValueError: if image.shape is not a [3] vector.
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: `image.shape`

File: tensorflow/python/ops/image_ops.py [hunk 11]
```diff
@@ -212,16 +247,26 @@ def _Check3DImage(image, require_static=True):
...
+
+  Returns:
+    An empty list, if image has fully defined dimensions. Otherwise, a list
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: `image`

File: tensorflow/python/ops/image_ops.py [hunk 12]
```diff
@@ -484,24 +532,38 @@ def crop_to_bounding_box(image, offset_height, offset_width, target_height,
...
+                        'height must be >= target + offset.')
+  image = control_flow_ops.with_dependencies(assert_ops, image)
+
+  cropped = array_ops.slice(
+    image,
+    array_ops.pack([offset_height, offset_width, 0]),
+    array_ops.pack([target_height, target_width, -1]))
+
+  cropped_shape = [None if is_tensor(i) else i
+                  for i in [target_height, target_width, depth]]
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: misalignment

File: tensorflow/python/ops/image_ops.py [hunk 13]
```diff
@@ -531,43 +593,74 @@ def resize_image_with_crop_or_pad(image, target_height, target_width):
...
+  assert_ops = []
+  assert_ops += _Check3DImage(image, require_static=False)
+  assert_ops += _assert(target_width > 0, ValueError,
+                        'target_width must be > 0.')
+  assert_ops += _assert(target_height > 0, ValueError,
+                        'target_height must be > 0.')
+
+  image = control_flow_ops.with_dependencies(assert_ops, image)
+  # `crop_to_bounding_box` and `pad_to_bounding_box` have their own checks.
+  # Make sure our checks comes first, so that error messages are clearer.
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: \# Make sure our checks **come** first ...

File: tensorflow/python/ops/image_ops.py [hunk 14]
```diff
@@ -531,43 +593,74 @@ def resize_image_with_crop_or_pad(image, target_height, target_width):
...
+  if is_tensor(target_height):
+    target_height = control_flow_ops.with_dependencies(
+      assert_ops, target_height)
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops.py]: line wrap not necessary

File: tensorflow/python/ops/image_ops_test.py [hunk 15]
```diff
@@ -550,45 +655,148 @@ def testError(self):
...
+
+    # Each line is a test configuration:
+    #   (offset_height, offset_width, target_height, target_width), err_msg
+    test_config = (([-1, 0, 4, 4], 'offset_height must be >= 0'),
+                   ([0, -1, 4, 4], 'offset_width must be >= 0'),
+                   ([2,  0, 4, 4], 'height must be <= target - offset'),
+                   ([0,  2, 4, 4], 'width must be <= target - offset'))
+
+    for params, err_msg in test_config:
+      self._assertRaises(x, x_shape, *params, err_msg)
```
Comment (gaohuazuo) [inline tensorflow/python/ops/image_ops_test.py]: python 2 compatibility

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
There are some failing tests in this, can you check?

**Comment by gaohuazuo** [comment]
@martinwicke `//tensorflow/python/image_ops_test` and `//tensorflow/contrib/framework:tensor_util_test` fixed. There are other failing tests, such as `//tensorflow/core/platform/cloud:oauth_client_test`, but I believe they are irrelevant to the PR.

**Comment by martinwicke** [comment]
Sorry, apparently we accumulated conflicts. Can you rebase?

**Comment by gaohuazuo** [comment]
@martinwicke The conflict was caused by `import` statements inserted into the same location. Fixed.

**Comment by rmlarsen** [comment]
It looks like there are still conflicts. Can you check?

**Comment by gaohuazuo** [comment]
Conflicts resolved.

**Comment by martinwicke** [comment]
Jenkins, test this please

**Comment by martinwicke** [comment]
Excellent! Thank you!

**Pull request merged**